satpy/tests/reader_tests/test_seviri_l2_bufr.py

```diff
@@ -97,20 +172,46 @@ def seviri_l2_bufr_test(self, filename):
                     ec2.return_value = 1
                     with mock.patch('eccodes.codes_release') as ec5:
                         ec5.return_value = 1
-                        z = fh.get_dataset(None, DATASET_INFO)
-                        # concatenate the original test arrays as
-                        # get dataset will have read and concatented the data
-                        x1 = np.concatenate((samp1, samp1), axis=0)
-                        np.testing.assert_array_equal(z.values, x1)
-                        self.assertEqual(z.attrs['platform_name'],
-                                         DATASET_ATTRS['platform_name'])
-                        self.assertEqual(z.attrs['ssp_lon'],
-                                         DATASET_ATTRS['ssp_lon'])
-                        self.assertEqual(z.attrs['seg_size'],
-                                         DATASET_ATTRS['seg_size'])
+                        z = fh.get_dataset(make_dataid(name='dummmy', resolution=48000), dataset_info)
+
+        return z
+
+    def as_swath_definition(self, fh, z, samp1):
+        """Perform checks if data loaded as swath definition."""
+        # With swath definition there will be no AreaDefinition implemented
+        self.assertRaises(NotImplementedError, fh.get_area_def, None)
+        # concatenate original test arrays as get_dataset will have read and concatented the data
+        x1 = np.concatenate((samp1, samp1), axis=0)
+        np.testing.assert_array_equal(z.values, x1)
+
+    def as_area_definition(self, fh, z, samp1):
+        """Perform checks if data loaded as AreaDefinition."""
+        ad = fh.get_area_def(None)
+        self.assertEqual(ad, AREA_DEF)
+
+        # Put BUFR data on 2D grid that the 2D array returned by get_dataset should correspond to
+        icol, irow = ad.get_array_coordinates_from_lonlat(fh.longitude.compute(),
+                                                          fh.latitude.compute())
+        icol, irow = np.ceil(icol).astype(int), np.ceil(irow).astype(int)
+
+        data = np.empty(ad.shape)
+        data[:] = np.nan
+        data[irow, icol] = np.concatenate((samp1, samp1), axis=0)
+        np.testing.assert_array_equal(z.values, data)
+
+        # Test that the correct AreaDefinition is identified for products with 3 pixel segements
+        fh.seg_size = 3
+        ad_ext = fh._construct_area_def(make_dataid(name='dummmy', resolution=9000))
+        self.assertEqual(ad_ext, AREA_DEF_EXT)
 
     def test_seviri_l2_bufr(self):
         """Call the test function."""
-        self.seviri_l2_bufr_test('GIIBUFRProduct_20191106130000Z_00_OMPEFS04_MET11_FES_E0000')
-        self.seviri_l2_bufr_test('MSG4-SEVI-MSGGIIN-0101-0101-20191106130000.000000000Z-20191106131702-1362128.bfr')
-        self.seviri_l2_bufr_test('MSG4-SEVI-MSGGIIN-0101-0101-20191106101500.000000000Z-20191106103218-1362148')
+        self.seviri_l2_bufr_test('ASRBUFRProd_20191106130000Z_00_OMPEFS01_MET08_FES_E0000')
```

Comment:
This test would be good as a pytest test with `@pytest.mark.parametrize` for each of the filenames and as_area_def options.

Reply:
I'm not so used to pytest (and unit testing in general), can please elaborate on this?

Reply:
So right now this test class is a subclass of the standard library's `unittest.TestCase` (or whatever it is) class. We run our tests with `pytest` now so we don't *need* to use `unittest` and can create regular functions (ex. `def test_X`) or test classes (ex. `TestSomething`) with test methods (ex. `def test_something_basic(self)`). When we don't use unittest classes we can use more functionality provided by pytest like the parametrize decorator: https://docs.pytest.org/en/6.2.x/parametrize.html

So if you moved this method outside the unittest class, you could do:

```python
@pytest.mark.parametrize(input_file,
[
    "GIIBUFRProduct_20191106130000Z_00_OMPEFS04_MET11_FES_E0000",
    ... other filenames ...,
])
def test_seviri_l2_bufr(self, input_file):
    seviri_l2_bufr_test(input_file)
```

I realize though that the `sevirir_l2_bufr_test` is a method on the unittest class so that would have to be extracted and changed to use regular assert statements (ex. `assert X == Y` instead of `self.assertEqual(X, Y)`). Not sure how much work you want to do on this. It could result in a rewrite of the unittest class.

Reply:
I did try your proposal, you can have a look and see if it's fine like that. I agree that it looks a bit cleaner!